lib/kaffe/producer.ex

```diff
@@ -148,16 +167,25 @@ defmodule Kaffe.Producer do
 
   defp add_timestamp(messages) do
     messages
-    |> Enum.map(fn {key, message} ->
-      {System.system_time(:millisecond), key, message}
-    end)
+    |> Enum.map(&add_timestamp_to_message/1)
   end
 
+  defp add_timestamp_to_message(message) when is_map(message) and not :erlang.is_map_key(:ts, message),
+    do: message |> Map.put(:ts, System.system_time(:millisecond))
+
+  defp add_timestamp_to_message(message) when is_map(message), do: message
```

Comment:
Ah, is this case to handle messages that already have a timestamp?

Reply:
Yes indeed. This allows users to set their own custom timestamp using the `ts` key, which is the same key that is used by `:brod`. If it is not set it will be set automatically.